### PR TITLE
Feature/documents

### PR DESCRIPTION
**Pull request by tsubik**
The first iteration of documents upload added to Litigations

Document could be
- an uploaded one
- external (we only store URL)

TODO
- we need to add google storage service later 
- could the same document be attached to many different records? The third option "Select existing" would be needed if that is the case.
- language column is missing, will be added in another story.

upload-documents-external-links

### CODE REVIEW

File: app/models/legislation.rb [hunk 1]
```diff
@@ -1,3 +1,18 @@
+# == Schema Information
+#
+# Table name: legislations
```
Comment (kowal) [inline app/models/legislation.rb]: I think you need to rebase against `develop`. Not sure why github is showing those changes.

Comment (tsubik) [reply app/models/legislation.rb]: I think that is what I did.

Comment (tsubik) [reply app/models/legislation.rb]: or maybe not...

Comment (tsubik) [reply app/models/legislation.rb]: done :+1:

File: app/decorators/litigation_decorator.rb [hunk 2]
```diff
@@ -16,4 +16,12 @@ def summary
   def core_objective
     model.core_objective.html_safe
   end
+
+  def document_list
```
[the remark this answers is not in the document]
Comment (tsubik) [reply app/decorators/litigation_decorator.rb]: good idea.

File: app/decorators/document_decorator.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+class DocumentDecorator < Draper::Decorator
+  delegate_all
+
+  def name_link
```
Comment (kowal) [inline app/decorators/document_decorator.rb]: what do you think about renaming those methods ..
`name_link` => `document_page_link` (or `just document_link`)
`open_link` => `document_url_link`
.. to indicate where those links actually go.

File: app/validators/url_validator.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
+class UrlValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    return unless value.present?
```
Comment (kowal) [inline app/validators/url_validator.rb]: could we combine `return unless value.present?` and `..unless url_valid?(value)` in single condition?

File: app/validators/url_validator.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+    record.errors[attribute] << (options[:message] || 'must be a valid URL') unless url_valid?(value)
+  end
+
+  def url_valid?(url)
+    url = begin
```
Comment (kowal) [inline app/validators/url_validator.rb]: I think `url = URI.parse(url) rescue false` would be enough here

File: app/javascript/controllers/document_list_controller.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import { Controller } from 'stimulus';
+
+export default class extends Controller {
+  static targets = [ 'links', 'templateExternal', 'templateUploaded']
+
+  connect() {
+    this.wrapperClass = this.data.get('wrapperClass') || 'nested-fields';
+  }
+
+  addRecord(event) {
```
Comment (kowal) [inline app/javascript/controllers/document_list_controller.js]: This code here looks like it shouldn't be directly related with "documents", rather, it is more like generic add/remove resource logic, right?

File: app/validators/url_validator.rb (hunk 5 above)
Comment (tsubik) [reply app/validators/url_validator.rb]: Rubocop changed that automagically.

File: app/javascript/controllers/document_list_controller.js (hunk 6 above)
Comment (tsubik) [reply app/javascript/controllers/document_list_controller.js]: You are right. I will change that

File: app/validators/url_validator.rb (hunk 4 above)
Comment (tsubik) [reply app/validators/url_validator.rb]: Sure, I will make two separate guard clauses. I think 2 will be more readable.

File: app/validators/url_validator.rb (hunk 5 above)
Comment (kowal) [reply app/validators/url_validator.rb]: ok, let's not mess with RuboCop 🤖

Comment (tsubik) [reply app/validators/url_validator.rb]: I already did. lol. I don't like that rule

Comment (kowal) [reply app/validators/url_validator.rb]: haha ok )

**Pull request merged**